k8s-openapi-codegen-common/src/templates/schema.rs

```diff
@@ -0,0 +1,17 @@
+pub(crate) fn generate(
+	mut writer: impl std::io::Write,
+	type_name: &str,
+	schema: &str,
+) -> Result<(), crate::Error> {
+	let lines = schema[2..schema.len()-1].lines().map(|s| format!("        {}", s)).collect::<Vec<_>>().join("\n");
+	writeln!(writer)?;
+	writeln!(writer, r#"#[cfg(feature = "schema")]"#)?;
+	writeln!(writer, r#"impl {type_name} {{"#, type_name = type_name)?;
+	writeln!(writer, r"    pub fn schema() -> serde_json::Value {{")?;
```

Comment:
If you're going to implement the thing for expanding `$ref`s anyway, is there a reason to return `serde_json::Value` here instead of `schemars::JsonSchema` ?

Also, that would allow us to emit code that directly creates a `schemars::SchemaObject`, instead of round-tripping through `serde_json::json!`. Then the user doesn't have to compile every invocation of that proc macro.

Reply:
Do you mean, why not `impl schemars::JsonSchema`?

1. Making `k8s_openapi` depend on `schemars` felt awkward to me. I like how it doesn't restrict users to a specific implementation. `kube` _currently_ uses `schemars`, but there are some missing features that might not make sense to add to a general JSON schema library, so this might change in the future. By keeping the schema `serde_json::Value`, it can be also useful for other use cases (as you wrote in https://github.com/Arnavion/k8s-openapi/issues/86#issuecomment-761106078). Also, it should be pretty easy to make use of this information for `schemars`.
2. `impl schemars::JsonSchema` seems like a lot of work if we want to avoid `serde_json::json!`. Methods used in derived `JsonSchema` to help are private too. Maybe I'm missing something, though. I haven't manually implemented `JsonSchema` for complex types. We'll need to generate code to build `SchemaObject` from `swagger20::Schema`. For each `$ref`, we'll need to do `<T as schemars::JsonSchema>::add_schema_as_property`, but this is not part of the public API.

Reply:
If `#[derive(schemars::JsonSchema]` can do it, then it's public API. Unless it's `#[doc(hidden)]` because the maintainer doesn't want other people to use it. Looking at what `#[derives(schemars::JsonSchema)]` expands to should answer that.

Reply:
It's `#[doc(hidden)]` in the latest published version, and it was then removed (https://github.com/GREsau/schemars/commit/55b860428e4526ef34886cb877af58256de6b046#diff-88895b0c150fd6f777b3df6591cbb00c7b88b100f8d99b769babe68315fd2cd0).

> How does a user who does want to use `#[derive(schemars::JsonSchema)]` on their type that contains `k8s_openapi` types use this `serde_json::Value` ?

The easiest is by using `#[schemars(schema_with = "schema_fn")]`, which is already necessary when adding validations and Kubernetes extensions. The original example from #86 is possible with:
```rust
#[derive(CustomResource, Debug, Clone, Deserialize, Serialize, JsonSchema)]
#[kube(group = "example.com", version = "v1", kind = "Example")]
#[kube(shortname = "ex", namespaced)]
pub struct ExampleSpec {
    #[schemars(schema_with = "optional_pvc_spec_schema")]
    pub pvcspec: Option<PersistentVolumeClaimSpec>,
}

fn optional_pvc_spec_schema(_: &mut schemars::gen::SchemaGenerator) -> schemars::schema::Schema {
    let mut schema = PersistentVolumeClaimSpec::schema();
    // for `Option`
    schema
        .as_object_mut()
        .expect("schema object")
        .insert("nullable".to_owned(), serde_json::json!(true));
    serde_json::from_value(schema).unwrap()
}
```

If we add `k8s_openapi::Schema` trait, we can do something like this too:

```rust
#[derive(CustomResource, Debug, Clone, Deserialize, Serialize, JsonSchema)]
#[kube(group = "example.com", version = "v1", kind = "Example")]
#[kube(shortname = "ex", namespaced)]
pub struct ExampleSpec {
    #[schemars(with = "Nullable<PersistentVolumeClaimSpec>")]
    pub pvcspec: Option<PersistentVolumeClaimSpec>,
}
```
with
```rust
struct Nullable<'a, T> {
    phantom: std::marker::PhantomData<&'a T>,
}

impl<T> schemars::JsonSchema for Nullable<'_, T>
where
    T: k8s_openapi::Schema,
{
    fn schema_name() -> String {
        // Doesn't guarantee uniqueness, but it doesn't matter for us because
        // we're not using `definitions`.
        std::any::type_name::<T>().to_owned()
    }

    fn json_schema(_: &mut schemars::gen::SchemaGenerator) -> schemars::schema::Schema {
        let mut schema = T::schema();
        schema
            .as_object_mut()
            .expect("schema object")
            .insert("nullable".to_owned(), serde_json::json!(true));
        serde_json::from_value(schema).unwrap()
    }
}
```

Reply:
I'll add the trait because it can be just:

```rust
pub struct ExampleSpec {
    #[schemars(with = "Option<SchemaValue<PersistentVolumeClaimSpec>>")]
    pub pvcspec: Option<PersistentVolumeClaimSpec>,
}
```